src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -195,14 +221,23 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
         private void CaptureVariable(Symbol variable, SyntaxNode syntax)
         {
             var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
-            if (type.IsRestrictedType())
+            if (type.IsRestrictedType() ||
+                (variable is LocalSymbol { RefKind: not RefKind.None } refLocal && !canRefLocalBeHoisted(refLocal)))
             {
                 (_lazyDisallowedCaptures ??= new MultiDictionary<Symbol, SyntaxNode>()).Add(variable, syntax);
             }
             else
             {
                 if (_variablesToHoist.Add(variable) && variable is LocalSymbol local && _boundRefLocalInitializers.TryGetValue(local, out var variableInitializer))
-                    CaptureRefInitializer(variableInitializer, syntax);
+                    CaptureRefInitializer(variableInitializer, local.SynthesizedKind != SynthesizedLocalKind.UserDefined ? variableInitializer.Syntax : syntax);
```

Comment:
>local.SynthesizedKind != SynthesizedLocalKind.UserDefined ? variableInitializer.Syntax [](http://example.com/codeflow?start=63&length=86)

Could you point to a test that observes this behavior change?

Reply:
For example `ForEach_InIterator_13` would report the diagnostic at the whole foreach without the change:

```cs
                // (19,9): error CS9217: A 'ref' local cannot be preserved across 'await' or 'yield' boundary.
                //         foreach (ref int y in buffer)
                Diagnostic(ErrorCode.ERR_RefLocalAcrossAwait, @"foreach (ref int y in buffer)
        {
            y *= y;
            System.Console.Write(y);
            yield return 1;
        }").WithLocation(19, 9)
```